### PR TITLE
WIP: Add DeDoDe detector+descriptor

### PR DESCRIPTION
**Pull request by Parskatt**
#### Changes

Adds dedode: https://github.com/parskatt/dedode 

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by Parskatt** [comment]
@ducha-aiki made some start at adding dedode, let me know issues. I'll fix the obvious linter errors and so on.

**Comment by ducha-aiki** [comment]
Looks good overall. Could you please add a normalization function to the models, so the input image will be in [0,1] range, and you do the ImageNet normalization internally?

**Review by edgarriba — CHANGES_REQUESTED** [review]
Thanks for the PR ! i left a bunch of comments specially because we are adding a large codebase here and potential things to be abstracted later to reuse existing pieces inside kornia. I leave the final review to @johnnv1

File: kornia/feature/dedode/decoder.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            bias=bias,
+        )
+        norm = norm_type(out_dim) if norm_type is nn.BatchNorm2d else norm_type(num_channels=out_dim)
+        relu = nn.ReLU(inplace=True)
+        conv2 = nn.Conv2d(out_dim, out_dim, 1, 1, 0)
+        return nn.Sequential(conv1, norm, relu, conv2)
+
+    def forward(self, feats):
+        b, c, hs, ws = feats.shape
+        with torch.autocast("cuda", enabled=self.amp, dtype=self.amp_dtype):
```
[the remark this answers is not in the document]
Comment (Parskatt) [reply kornia/feature/dedode/decoder.py:87]: I had issues that autocasting to float16 doesnt work for cpu, which I think it defaults to now, need to check.

File: kornia/feature/dedode/dedode.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    @torch.inference_mode()
+    def describe(self, images, keypoints=None):
+        self.train(False)
+        B, C, H, W = images.shape
+        if C != 3:
+            raise ValueError(f"DeDoDe expects RGB, got {C=}")
+        descriptions = self.descriptor.forward(images)
+        if keypoints is not None:
+            described_keypoints = F.grid_sample(
+                descriptions.float(), keypoints[:, None], mode="bilinear", align_corners=False
```
[the remark this answers is not in the document]
Comment (Parskatt) [reply kornia/feature/dedode/dedode.py:64]: This could have issues if images are in float16, I want specifically float32 here

File: kornia/feature/dedode/detector.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        self,
+        images,
+    ):
+        features, sizes = self.encoder(images)
+        logits = 0
+        context = None
+        scales = ["8", "4", "2", "1"]
+        for idx, (feature_map, scale) in enumerate(zip(reversed(features), scales)):
+            delta_logits, context = self.decoder(feature_map, context=context, scale=scale)
+            logits = logits + delta_logits.float()  # ensure float (need bf16 doesnt have f.interpolate)
```
[the remark this answers is not in the document]
Comment (Parskatt) [reply kornia/feature/dedode/detector.py:21]: Want specifically float32

File: test/feature/test_dedode.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        keypoints, scores, descriptions = dedode(inp, n=n)
+        assert keypoints.shape == (shape[0], n, 2)
+        assert scores.shape == (shape[0], n)
+        assert descriptions.shape == (shape[0], n, 256)
+
+    @pytest.mark.slow
+    def load_weights(self, dtype, device):
+        # only testing "B" as dinov2 is quite heavy
+        dedode = DeDoDe(descriptor_model="B").to(device, dtype)
+        dedode = DeDoDe(descriptor_model="G").to(device, dtype)
```
[the remark this answers is not in the document]
Comment (Parskatt) [reply test/feature/test_dedode.py:23]: Sounds good.

File: kornia/feature/dedode/decoder.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        self,
+        in_dim=6,
+        hidden_dim=16,
+        out_dim=2,
+        dw=True,
+        kernel_size=5,
+        hidden_blocks=5,
+        amp=True,
+        residual=False,
+        amp_dtype=torch.float16,
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/feature/dedode/decoder.py:32]: I believe that it is better to have a fix and at leaset theoretical ability to run anywhere, at least for debug purposes. I love my macbook :)

File: kornia/feature/dedode/transformer/layers/mlp.py
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/feature/dedode/transformer/layers/mlp.py:1]: @edgarriba Yes, but only if that doen't affect the model itself.

File: test/feature/test_dedode.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply test/feature/test_dedode.py:23]: You can use DISK tests as an example - run on some small images with low amount of keypoints.
https://github.com/kornia/kornia/blob/2a56ccd1fde734848cd94b38d06d14ef6e765e6a/tests/feature/test_disk.py#L43

File: kornia/feature/dedode/dedode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/feature/dedode/dedode.py:64]: @edgarriba no, we can do a workaround

**Comment by ducha-aiki** [comment]
Closing in favor of https://github.com/kornia/kornia/pull/2835
